### DISCUSSION TITLE
TltxY and TltY not working on X12s

### DISCUSSION THREAD

**Discussion by Kevltan** [EdgeTX/edgetx#2771] [General]
********************************
Solution to the IMU issue at the end of these comments.
********************************

TltX and TltY not working. Works fine after firmware install on exiting (without power off TX). But when powered off and back on again the values are frozen and no wake up on screen dimming. Any ideas anybody please? My Tx has the touch screen mod which works fine.

**Comment by pfeerick** [comment]
So, for context, are you "KevTheRev" on Discord? Meaning that you set `option(IMU_LSM6DS33 "Enable I2C2 and LSM6DS33 IMU" ON)` and it's still not working for you?

**Comment by Kevltan** [comment]
No still does not work... works sometimes which is what I can't understand
there is no logic other than when it works after first loading firmware and
exiting. Seemed to work for ages after doing this mod above and then just
stopped working again!

**Comment by Kevltan** [comment]
There must be something in the startup when I power up, a firmware reload
always clears the problem until power is re-cycled. Where upon it can work
for a bit or can fail straight away. I have loaded basic model setups
before powering off and powering back on thinking this may be a problem but
this makes no difference.

**Comment by Kevltan** [comment]
I have re-routed the I2C bus cable again that goes to the touch screen
modification. I believe (although I am not absolutely sure) that I was
getting "noise" from the RF module as I was getting some very strange
movements on L1, RL pots, they were jittering on the screen and moving
unpredictably. When I re-routed the individual twisted wires differently
the problem went away. So I then decided to use 3.5mm screened 2 core cable
the outer screen conductor soldered to -0V and routed this above the RF
module as I did  with the 2 twisted individual I2C wires this cured the
L1, RL pot jitter described above. I have now re-routed the screened cable
below the RF module and across the back of pcb over the L1, L2 sliders
again, trying to take the shortest route. Since I have done this the IMU
"seems" to be working again and sliders are stable. I am not "counting my
chickens before they have hatched" but this might be the variable that
l/we've been looking for and may be a consequence of this mod!! This would
also account for the unpredictability (works then doesnt work). I will test
and keep you updated. 👍🙂
What I have also noticed is that in screen settings/setup widgets, that if
certain values are changed or widgets removed, including TltX and TltY, that
the IMU sometimes but not always freezes until a power off/power on is
performed - is this normal?

**Comment by Kevltan** [comment]
![20221120_195501](https://user-images.githubusercontent.com/56512603/202924289-213635c5-53c6-4a01-876d-130d01cd8e74.jpg)
This is the new cable routing arangement. Same as my original 2 twisted pair wire routing but this time screened and doesn't seem to effect
L1 and RL pots on the touch screen anymore.

**Reply by Kevltan** [reply]
This has sorted the problem I have had no more power up freezes of the IMU......
 I had also noted (since I did the touch mod) that telemetry for my turbine had been difficult to get to bind on switch on, it took a few attempts of connecting and disconnecting receiver and repositioning of transmitter. The cable re-route has sorted this - no more telemetry connection problems! 
 In my findings so far the I2C SDA/SCL cable routing has to be as short as possible, screened and as far away from the RF module (or other forms of induced "noise") as possible. Same length cable run as picture below but running as close to the 0V PCB ground plane as possible and below RF module, close to main battery , whilst trying to avoid on board components.

**Reply by pfeerick** [reply]
Fingers crossed this has completely fixed the issue... shielding will certainly not have any downsides here as it is a noisy electrical and RF environment in there. 😄  I2C can go reasonable distances (1-10M without repeaters), but the other thing to check then would be whether the I2C bus has the correct pullup resistance - which would take some reverse-engineering to figure out. I presume the mainboard would have pullups, and then the two different I2C boards (touchpanel, IMU) might have them also, so there could also be an issue where the I2C bus is being pulled up too hard.

**Reply by Kevltan** [reply]
Thanks Peter, yes was thinking of pullup resistors and might be the consequence of this wire run across the main board but then was thinking its such a short cable run and the touch interface uses the onboard 3.3v supply, so logic levels should remain pretty consistent. Like you said this would be a very difficult thing to work out. I was worried when I did the mod that there maybe some RF issues with the cable run accross the board. it seems to behaving itself now though. 
When editing the widgets is it normal for the IMU to "hang" "freeze" until a power re-cycle?
It does not do it everytime and can't get to fail in normal use now.

**Reply by Kevltan** [reply]
The TltX and TltY problem is back again!
What I have noticed now is it wont come on normal power up, but if I hold the T1 T4 trims in and and power on as though I was going to do a firmware update and then exit instead, it works every time! Any ideas? I really am giving up here because there seems to be no logic in it. I think I've fixed and then I haven't! Stange thing is I was adding sound files when it stopped working again🤔

**Comment by Kevltan** [comment]
![20221120_152504](https://user-images.githubusercontent.com/56512603/202926682-8e18c63f-eaa3-4b37-8005-25d127762c9a.jpg) This is the routing of the original 2 twisted pair wires which I re-routed to here after having the pot jittering problem on the touch screen. I wasnt using the IMU then so don't know if it was effected by running so close the RF module. I am believing this to be the possible cause now and the poor telemetry connectivety.

**Comment by Kevltan** [comment]
![20221120_183046](https://user-images.githubusercontent.com/56512603/202927814-2c2ac58d-f1a8-4933-ac15-df1a42c0550b.jpg)
![20221120_160756](https://user-images.githubusercontent.com/56512603/202927819-4c57fad4-ebd1-41f4-b3ab-b1ed322a79c8.jpg) Pic1. Type of 3.5mm screened cable used. 
Pic2. Now stable and functioning IMU values TlltX & TltY on the Horus X12S.

**Comment by Luznatural** [comment]
Thanks Kevin for the information. Good job on the pictures.
I have received my parts, and as soon as I get the parts 3D printed I will do the mod.
Thanks again

**Comment by Luznatural** [comment]
Are you going to share the information with the guy that invented the mod? perhaps is a good idea to help more people, if he puts your information in his descrption.

**Comment by Kevltan** [comment]
Yes I will. The only reason I found these problems (apart from the L2/RL
pot jittering) was I wanted the screen to wake up on tilt of the radio and
I had noticed it was no longer functioning properly.
I had also noticed since the mod that telemetry had been affected as
described in my findings. Not forgetting that this had all been tested in a
"busy" 2.4Ghz/5Ghz WIFI environment in my home.
I have yet to test in the open field. But I think if it works here it's
going to work in lots of open space!
Cable routing definitely seems to be critical here if you want all
functions to work correctly.

**Comment by Kevltan** [comment]
The TltX and TltY problem is back again!
What I have noticed now is it wont come on normal power up, but if I hold the T1 T4 trims in and and power on as though I was going to do a firmware update and then exit instead, it works every time! Anybody any ideas? I really am giving up here because there seems to be no logic in it. I think I've fixed and then I haven't! Stange thing is I was adding sound files when it stopped working again. Still thinking there must be something in the startup/powerup routine that is different to the boot loader and exit routine - works everytime from the bootloader🤔.

**Comment by Kevltan** [comment]
All problems with IMU not working on start up caused by having a splash.png file too large in the System/Images file on the SD card. The size was 220kb this is too large and seems to upset the IMU on start up. I used an online .png file compressor to reduce it to 89kb without changing pixel count of 480 X 272 https://tiny-img.com/png-compressor/. I have tried this on cards from 256Mb - 32Gb standard and SDHC format. All formatted to Fat32 default allocation. IMU has not failed since! So if your running your own splash screen at start up it can't be much over 100kb in size. The ones I created online in https://www.air-rc.com/ or https://www.skyraccoon.com/ were too big and needed reducing.
Hope this helps anybody experiencing this problem.
